torchtitan/models/deepseek_v3/model/state_dict_adapter.py

```diff
@@ -190,20 +490,27 @@ def from_hf(self, hf_state_dict: dict[str, Any]) -> dict[str, Any]:
             if "mlp.experts" in key:
                 abstract_key = re.sub(r"(\d+)", "{}", key, count=2)
                 layer_num, expert_num = re.findall(r"\d+", key)
-                new_key = self.from_hf_map[abstract_key]
-                new_key = new_key.format(layer_num)
+                titan_abstract_key = self.from_hf_map[abstract_key]
+                new_key = titan_abstract_key.format(layer_num)
 
                 # Store the expert's weight in expert_weights_by_layer for concatenating later.
                 if layer_num not in expert_weights_by_layer:
                     expert_weights_by_layer[layer_num] = {}
-                if abstract_key not in expert_weights_by_layer[layer_num]:
-                    expert_weights_by_layer[layer_num][abstract_key] = {}
-                expert_weights_by_layer[layer_num][abstract_key][expert_num] = value
+                if titan_abstract_key not in expert_weights_by_layer[layer_num]:
+                    expert_weights_by_layer[layer_num][titan_abstract_key] = {}
+                expert_weights_by_layer[layer_num][titan_abstract_key][
+                    expert_num
+                ] = value
+
+                if isinstance(value, DTensor):
+                    stacked_value = self._concatenate_local_expert_weights(
```

Comment:
If we pass `layer_num` into `_concatenate_local_expert_weights`, can we avoid looping layers inside `_concatenate_local_expert_weights`?

Reply:
Oh yes, this is correct, let me update this part